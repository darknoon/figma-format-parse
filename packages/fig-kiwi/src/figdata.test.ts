import FigmaArchiveParser, { FigmaArchiveWriter } from "./archive";
import { readFileSync, writeFileSync } from "fs";
import {
  compileSchema,
  decodeBinarySchema,
  compileSchemaTypeScript,
  prettyPrintSchema,
  encodeBinarySchema,
} from "kiwi-schema";
import schema from "./schema";
import {
  Schema as CompiledSchema,
  NodeChange,
  Message,
  SparseMessage,
} from "./schema-defs";
import { deflateRaw, inflateRaw } from "pako";
import { readFigFile } from "./index";

test.skip("this just formats the schema", () => {
  const prettySchema = prettyPrintSchema(schema);
  writeFileSync(__dirname + "/figma.kiwi", prettySchema);
});

test("able to parse figma kiwi", () => {
  const data = readFileSync(__dirname + "/../data/blue-circle.fig");
  const parsed = FigmaArchiveParser.parseArchive(data);
  expect(parsed.header.version).toBe(15);
  const [schemaFile, dataFile, previewFile] = parsed.files;
  const figDataDec = inflateRaw(dataFile);
  expect(figDataDec.length).toBeGreaterThan(100);
  writeFileSync(__dirname + "/../data/blue-circle.fig-inflated", figDataDec);
  const fileSchema = decodeBinarySchema(inflateRaw(schemaFile));
  expect(fileSchema).toHaveProperty("definitions");
  const compiledSchema = compileSchema(fileSchema) as CompiledSchema;
  const decoded = compiledSchema.decodeMessage(figDataDec);
  expect(decoded).not.toBeNull();
});

test("compare red and blue", () => {
  const blue = readFigFile(
    readFileSync(__dirname + "/../data/blue-circle.fig")
  ).message;
  const red = readFigFile(
    readFileSync(__dirname + "/../data/red-circle.fig")
  ).message;
  expect(blue).toMatchSnapshot();
});

test("able to enc dec a known message", () => {
  const cs = compileSchema(schema) as CompiledSchema;
  const message: NodeChange = { guid: { sessionID: 123, localID: 321 } };
  const data = cs.encodeNodeChange(message);
  const decoded = cs.decodeNodeChange(data);
  expect(decoded).toEqual(message);
});

test("it's compressed binary kiwi format in the metadata", () => {
  const meta = readFileSync(__dirname + "/../data/red-circle.fig-meta");
  const schemaBin = inflateRaw(meta);
  writeFileSync(__dirname + "/../data/red-circle.bkiwi", schemaBin);
  const figMetaSchema = decodeBinarySchema(schemaBin);
  writeFileSync(
    __dirname + "/../data/red-circle.kiwi",
    prettyPrintSchema(figMetaSchema)
  );

  // writeFileSync(
  //   __dirname + "/fig-kiwi.ts",
  //   compileSchemaTypeScript(figMetaSchema)
  // );
  // console.log(tsdef);
});

test("able to decode a real websocket message", () => {
  const cs = compileSchema(schema) as CompiledSchema;
  const dataCompressed = readFileSync(__dirname + "/../data/set-black.bin");
  const data = inflateRaw(dataCompressed);
  const decoded = cs.decodeMessage(data);
  expect(decoded).not.toBeNull();
  expect(decoded).toMatchSnapshot();
});

test("able to write dummy files to a fig-kiwi archive", () => {
  const encoder = new FigmaArchiveWriter();
  const expectedFiles = [
    new Uint8Array([128, 1, 2, 3, 4, 5]),
    new Uint8Array([256, 5, 6, 7, 8, 23, 11]),
  ];
  encoder.files = expectedFiles;
  const archive = encoder.write();

  // Now decode and verify
  const { header, files } = FigmaArchiveParser.parseArchive(archive);
  expect(files).toEqual(expectedFiles);
  expect(header).toEqual(encoder.header);
});

test("able to write a Message to an archive", () => {
  const binSchema = encodeBinarySchema(schema);
  const compiledSchema = compileSchema(schema) as CompiledSchema;
  const message: SparseMessage = JSON.parse(
    readFileSync(__dirname + "/../data/grey-circle-paste.json", {
      encoding: "utf8",
    })
  );
  const encoder = new FigmaArchiveWriter();
  encoder.files = [
    deflateRaw(binSchema),
    deflateRaw(compiledSchema.encodeMessage(message)),
  ];
  const data = encoder.write();
  writeFileSync(__dirname + "/../data/grey-circle-paste-generated.fig", data);
});
