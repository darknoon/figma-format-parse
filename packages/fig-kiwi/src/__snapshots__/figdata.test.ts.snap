// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`able to decode a real websocket message 1`] = `
{
  "ackID": 265,
  "nodeChanges": [
    {
      "fillPaints": [
        {
          "blendMode": "NORMAL",
          "color": {
            "a": 1,
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "guid": {
        "localID": 5,
        "sessionID": 9,
      },
    },
  ],
  "sessionID": 9,
  "type": "NODE_CHANGES",
}
`;

exports[`compare red and blue 1`] = `
{
  "ackID": 0,
  "blobs": [
    {
      "bytes": Uint8Array [
        1,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        4,
        0,
        0,
        17,
        67,
        207,
        20,
        225,
        66,
        207,
        20,
        225,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        0,
        0,
        17,
        67,
        4,
        97,
        214,
        1,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        0,
        0,
        207,
        20,
        225,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        145,
        66,
        4,
        0,
        0,
        0,
        0,
        97,
        214,
        1,
        66,
        97,
        214,
        1,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        145,
        66,
        0,
        0,
        0,
        0,
        4,
        207,
        20,
        225,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        17,
        67,
        97,
        214,
        1,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        0,
      ],
    },
  ],
  "nodeChanges": [
    {
      "blendMode": "PASS_THROUGH",
      "guid": {
        "localID": 0,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Document",
      "opacity": 1,
      "phase": "CREATED",
      "transform": {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "DOCUMENT",
      "visible": true,
    },
    {
      "backgroundColor": {
        "a": 1,
        "b": 0.9607843160629272,
        "g": 0.9607843160629272,
        "r": 0.9607843160629272,
      },
      "backgroundEnabled": true,
      "backgroundOpacity": 1,
      "blendMode": "PASS_THROUGH",
      "guid": {
        "localID": 1,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Page 1",
      "opacity": 1,
      "parentIndex": {
        "guid": {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "transform": {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": true,
    },
    {
      "blendMode": "PASS_THROUGH",
      "guid": {
        "localID": 2,
        "sessionID": 0,
      },
      "internalOnly": true,
      "mask": false,
      "maskIsOutline": false,
      "name": "Internal Only Canvas",
      "opacity": 1,
      "parentIndex": {
        "guid": {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "~",
      },
      "phase": "CREATED",
      "transform": {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": false,
    },
    {
      "arcData": {
        "endingAngle": 6.2831854820251465,
        "innerRadius": 0,
        "startingAngle": 0,
      },
      "blendMode": "PASS_THROUGH",
      "cornerRadius": 0,
      "cornerSmoothing": 0,
      "dashPattern": [],
      "fillGeometry": [
        {
          "commandsBlob": 0,
          "styleID": 0,
          "windingRule": "NONZERO",
        },
      ],
      "fillPaints": [
        {
          "blendMode": "NORMAL",
          "color": {
            "a": 1,
            "b": 0.8500000238418579,
            "g": 0.6204999685287476,
            "r": 0.53125,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "guid": {
        "localID": 2,
        "sessionID": 2,
      },
      "handleMirroring": "NONE",
      "horizontalConstraint": "MIN",
      "locked": false,
      "mask": false,
      "maskIsOutline": false,
      "miterLimit": 4,
      "name": "Ellipse 1",
      "opacity": 1,
      "parentIndex": {
        "guid": {
          "localID": 1,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "proportionsConstrained": true,
      "size": {
        "x": 145,
        "y": 145,
      },
      "strokeAlign": "INSIDE",
      "strokeCap": "NONE",
      "strokeGeometry": [],
      "strokeJoin": "MITER",
      "strokePaints": [],
      "strokeWeight": 1,
      "transform": {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "ELLIPSE",
      "verticalConstraint": "MIN",
      "visible": true,
    },
  ],
  "sessionID": 0,
  "type": "NODE_CHANGES",
}
`;
